app/models/bulkrax/oai_entry.rb

```diff
@@ -44,17 +38,30 @@ def build_metadata
       end
       add_metadata('thumbnail_url', thumbnail_url)
 
+      parsed_metadata[Bulkrax.system_identifier_field] ||= [record.header.identifier]
+      parsed_metadata['contributing_institution'] = [contributing_institution]
+
       add_visibility
       add_rights_statement
-      self.parsed_metadata['source'] ||= [record.header.identifier]
-      self.parsed_metadata['contributing_institution'] = [contributing_institution]
+      add_collections
 
-      if collection.present?
-        self.parsed_metadata['collections'] ||= []
-        self.parsed_metadata['collections'] << {id: self.collection.id}
-      end
+      parsed_metadata
+    end
 
-      return self.parsed_metadata
+    # Retrieve list of collections for the entry; add to collection_ids
+    # If OAI-PMH doesn't return setSpec in the headers for GetRecord, use parser.collection_name
+    #   in this case, if 'All' is selected, records will not be added to a collection.
+    def collections
+      if sets.blank?
+        c = Collection.where(Bulkrax.system_identifier_field => parser.collection_name).first
+        collection_ids << c.id unless c.blank? || collection_ids.include?(c.id)
+      else
+        sets.each do |set|
+          c = Collection.where(Bulkrax.system_identifier_field => set.content).first
+          collection_ids << c.id unless c.blank? || collection_ids.include?(c.id)
+        end
+      end
+      @collections ||= collection_ids
```

Comment:
if you are going to ||= this, I think you should escape out of it at the top of the method. You're doing a lot of work you throw away if @collections is already set. 

Also this has the same "name is collections but is now returning collection_ids" problem mentioned above.

Reply:
I've renamed to `retrieve_or_create_collection_ids` ... a bit of a mouthful, but it's clearer I think